src/query/v2/physical/multiframe.hpp

```diff
@@ -0,0 +1,236 @@
+///
+/// At the moment OPTION 2 is implemented. Multiframe is not thread-safe, but
+/// access to the Multiframe reference is "protected" by appropriate methods.
+/// The intention is to make interactions with the Multiframe as fast as
+/// possible, while minimizing the overhead of concurrent access while getting
+/// correct Multiframe references.
+///
+/// TODO(gitbuda): Articulate how to actually do/pass aggregations in the Multiframe context.
+///
+/// Moving/copying Frames between operators is questionable because operators
+/// mostly operate on a single Frame value.
+///
+class Multiframe {
+ public:
+  Multiframe() = delete;
+  Multiframe(const Multiframe &) = default;
+  Multiframe(Multiframe &&) = default;
+  Multiframe &operator=(const Multiframe &) = default;
+  Multiframe &operator=(Multiframe &&) = default;
+  explicit Multiframe(const size_t capacity = 0) : data_(0) { data_.reserve(capacity); }
+  ~Multiframe() = default;
+
+  size_t Capacity() const { return data_.capacity(); }
+  size_t Size() const { return data_.size(); }
+  bool IsFull() const { return Size() == Capacity(); }
+  void PushBack(const Frame &frame) { data_.push_back(frame); }
+  void Put(const Frame &frame, int at) { data_[at] = frame; }
+  const std::vector<Frame> &Data() const { return data_; }
+
+ private:
+  std::vector<Frame> data_;
+};
+
+struct Token {
+  int id;
+  Multiframe *multiframe;
+};
+
+template <typename TPool>
+concept MultiframePoolConcept = requires(TPool p, int id) {
+  { p.GetEmpty() } -> std::same_as<std::optional<Token>>;
+  { p.GetFull() } -> std::same_as<std::optional<Token>>;
+  { p.ReturnEmpty(id) } -> std::same_as<void>;
+  { p.ReturnFull(id) } -> std::same_as<void>;
+};
+
+enum class Mode {
+  // TODO(gitbuda): To implement FCFS ordering policy a queue of token is
+  // required with round robin id assignment.
+  //   * A queue seem very intuitive choice but there are a lot of issues.
+  //   * Trivial queue doesn't work because returning token is not trivial.
+  // FCFS/LIFO is a probably a wrong concept here, instead:
+  //   * Single/Multiple Producer + Single/Multiple Consumer is probably the way to go.
+  //
+  SPSC,  // With 2 Multiframes SPSC already provides simultaneous operator execution.
+  SPMC,
+  MPSC,
+  MPMC,
+  // All the above 4 modes are meaningful in the overall operator stack. If
+  // MultiframePool is used between the operators, each operator can specify
+  // the actual mode because of the operator semantics.
+  //
+};
+enum class Order {
+  FCFS,
+  RANDOM,
+};
+
+enum class PoolState {
+  EMPTY,
+  HAS_MORE,
+  EXHAUSTED,
+};
+
+/// Preallocated memory for an operator results.
+/// Responsible only for synchronizing access to a set of Multiframes.
+/// Requires giving back Token after Multiframe is consumed/filled.
+/// Equivalent to a queue of Multiframes with intention to minimize the time
+/// spent in critical sections.
+///
+class MPMCMultiframeFCFSPool {
```

Comment:
Rename to `ResultPool` and templatize by `TResult` which is going to be the Multiframe